activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb

```diff
@@ -319,7 +338,9 @@ def discard! # :nodoc:
       end
 
       def native_database_types #:nodoc:
-        NATIVE_DATABASE_TYPES
+        types = NATIVE_DATABASE_TYPES.dup
+        types[:datetime] = types[datetime_type]
+        types
```

Comment:
Might be worth caching this - `native_database_types` is mainly (only?) called by SchemaDumper but it does it twice for every column dumped.

Reply:
Good idea - updated.

Reply:
Hmmm. The tests [didn't like that](https://buildkite.com/rails/rails/builds/77093#ea12f75d-abe4-4f02-aaf1-710a4735d707), I guess because they aren't reloading the adapter between each test. I'm going to revert for now and investigate if memoization is necessary - I don't think this is called much at runtime.

Reply:
> Hmmm. The tests [didn't like that](https://buildkite.com/rails/rails/builds/77093#ea12f75d-abe4-4f02-aaf1-710a4735d707), I guess because they aren't reloading the adapter between each test. I'm going to revert for now and investigate if memoization is necessary - I don't think this is called much at runtime.

Clearing the instance variable used for caching the results before/after the relevant tests should fix that

Reply:
Yep - that's what I ended up doing. This is what the changes ended up looking like: https://gist.github.com/ghiculescu/eb14ab894824d6914a1a2360d8b14937

It's fairly messy, I'd love any ideas on how to simplify.

Reply:
I think the implementation is fine, maybe just extract the method to a module and put it in `test/support` so we're not duplicating it.

Reply:
Yep that’s what I did in a subsequent commit. It’s now all a single helper method. 

Thanks for your help here! Did you have any other thoughts on this PR?

Reply:
Is the comment @matthewd made about changing existing columns in schema.rb been addresses - I think so but just want to confirm.

Reply:
Yes, I think so too. 

Some of the comments are stale since the direction of the PR changed a bit (eg the reference to using new framework defaults). The PR body is up to date though.

Reply:
> Yes, I think so too.

Can you confirm by running `rails db:schema:dump` on a PostgreSQL app with timestamp and timestamptz columns - once with `datetime_type` set to `:timestamp` and once with it set to `:timestamptz`. If there's no changes to the schema.rb file then we're good to go.

Reply:
You could add a railties test for this in: 
https://github.com/rails/rails/blob/main/railties/test/application/configuration_test.rb

Reply:
Yep, will do tomorrow and report back 👍

Reply:
> Is the comment @matthewd made about changing existing columns in schema.rb been addressed - I think so but just want to confirm.

What @matthewd said in https://github.com/rails/rails/pull/41084#issuecomment-776414131 is that the configureation should not change the behavior for all existing migrations. See #34970, #25395, #26266 for your reference.

Reply:
@pixeltrix  @kamipo thanks for those links. I had a look through them, and I think this PR is a bit different to those. In all those cases, we are changing how Rails handles certain commands in `db:migrate` without configuration.

- https://github.com/rails/rails/pull/34970 makes `t.timestamps` mean `t.timestamps null: false` from 6.0 onwards, but `t.timestamps null: true` before 6.0.
- https://github.com/rails/rails/pull/25395 makes `create_table(id: :uuid)` use a different UUID function from 5.1 onwards.
- https://github.com/rails/rails/pull/26266 makes `create_table` create bigint ID columns, for tables created on 5.1 onwards.

As I understand it, the point of all these is so that if you run an older migration, it still does the same thing. eg. if you run a migration created on Rails 5, it'll still use the old UUID function even if you have upgraded to 5.1.

(Sorry to spell out the obvious to you, but this is helping me confirm my understanding.)

The difference in this PR is that we are adding a new configuration as well. Nothing will change by default; to use it you have to do something like `ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.datetime_type = : timestamptz ` in an initializer.

Given that, I'm not sure there's a way to make it work with https://github.com/rails/rails/blob/main/activerecord/lib/active_record/migration/compatibility.rb because that would require something in either Rails or the user's code to indicate from which migration onward that setting was changed.

To work around this, I have added this warning to the PR body, release notes, and configuration guide:

> If you change this setting in an existing app, you should immediately call `bin/rails db:migrate` to ensure your `schema.rb` file remains correct.

If you do this, then with the above change, columns that were `t.datetime` will become `t.timestamp`, and `t.timestamptz` will become `t.datetime`. I am working on a test to demonstrate this.

Given that, I don't think any changes in https://github.com/rails/rails/blob/main/activerecord/lib/active_record/migration/compatibility.rb are required. Does this make sense, and am I on track?

Reply:
Isn't the difference is that in older migrations it should ignore the new config and always use `timestamp`?

Reply:
... 🤦 Yeah, I think so. Let me try that out.

Reply:
I can remove the comments and I will add tests, but is this what you had in mind? https://github.com/rails/rails/pull/41084/commits/c4290dae3da94b610a070b84377b94d201305465

Reply:
@ghiculescu yes, that's right - the inheritance chain is slightly confusing so I had think about it for a moment 🤯

Reply:
Haha, you're telling me! I think some comments about it for future developers would be helpful. Is there a preferred way to do that? I was thinking of adding a few lines explaining the inheritance chain here: https://github.com/rails/rails/blob/main/activerecord/lib/active_record/migration/compatibility.rb#L16 but I don't want them to end up as public documentation.